cmd/snap/cmd_model.go

```diff
@@ -0,0 +1,134 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/jessevdk/go-flags"
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/i18n"
+	yaml "gopkg.in/yaml.v2"
+)
+
+var (
+	shortModelHelp = i18n.G("Get the active model for this device")
+	longModelHelp  = i18n.G(`TODO`)
+)
+
+type cmdModel struct {
+	waitMixin
+	Serial    bool `long:"serial"`
+	Verbose   bool `long:"verbose"`
+	Assertion bool `long:"assertion"`
+}
+
+func init() {
+	cmd := addCommand("model",
+		shortModelHelp,
+		longModelHelp,
+		func() flags.Commander {
+			return &cmdModel{}
+		}, nil, []argDesc{},
+	)
+	cmd.hidden = true
+}
+
+func (x *cmdModel) Execute(args []string) error {
+	var assertion asserts.Assertion
+	var err error
+	if x.Serial {
+		assertion, err = x.client.CurrentSerialAssertion()
+	} else {
+		assertion, err = x.client.CurrentModelAssertion()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// --assertion means output the raw assertion
+	if x.Assertion {
+		_, err = Stdout.Write(asserts.Encode(assertion))
+		return err
+	}
+
+	w := tabwriter.NewWriter(Stdout, 2, 2, 1, ' ', 0)
+
+	primaryHeaderKeys := assertion.Type().PrimaryKey
+	sort.Strings(primaryHeaderKeys)
+
+	primaryHeaderKeyMap := make(map[string]bool)
+	for _, header := range primaryHeaderKeys {
+		primaryHeaderKeyMap[header] = true
+	}
+
+	for _, header := range primaryHeaderKeys {
+		val := assertion.HeaderString(header)
+		fmt.Fprintf(w, "%s:\t%s\n", header, val)
+	}
+
+	// --verbose means output all of the fields
+	if x.Verbose {
+		allHeaderMaps := assertion.Headers()
+		allHeaders := make([]string, len(allHeaderMaps))
+		i := 0
+		for k := range allHeaderMaps {
+			allHeaders[i] = k
+			i++
+		}
+
+		sort.Strings(allHeaders)
+		for _, headerName := range allHeaders {
+			header := allHeaderMaps[headerName]
+			if _, ok := primaryHeaderKeyMap[headerName]; ok {
+				continue
+			}
+			switch headerName {
+			case "authority-id", "format", "type":
+			default:
+				// handle different types of the header value
```

Comment:
doing this like this produces really nasty output, that is not readable, does not look at how wide things need to be to not wrap in nasty ways, and is not valid yaml.
If you had run it even once you would've noticed immediately. As you haven't, I'm stealing all your cookies for the day.

This is what it currently looks like, once I got the daemon to not panic:
```
~$ snap model --verbose --serial
brand-id:   generic
model:      generic-classic
serial:     c2f87913-59fe-4b50-a3b3-f1806f7187de
device-key: AcbBTQRWhcGAARAA1XsFG+t9RAqOnoc6IubGzilRR/to79ihTU9ukOnOhORKtKyZHo/2lRKYKr13
MoM3dWn1U7tKGHsrjjOUSYK5v4TTbXPt99bygfMGG4mi3kQ2/1x62UVGgfqKHNnN1jtKohAQVHsp
psr9/4+WVyvxRAX80jM5w7lo2I44lNITx9O6d1oFGWrxstPjtQJs5IlYRhnvzFDP6mMHLKZFtofJ
o62j+sgnxlWswTlv6Fn9yh6n6yFiPHc8y/bsNHPnWvG2s+e7fRFPFmN17ZEYiDB/1ob/IbWCk+yu
Tr8eM+SvvkL8I3fRKJgPHtEJ4e3RRCHO4s6FtSxn4x1Ny//tU0l2s71ykeQyi5nW/kw3Qr8ALggG
l9Bqt2lsOaFh5jNg1bF/O7ZaCD4SvNSejcXkT+kMhhq7N95JT4N1Wg67CBqtAi0Uvqf4F5U2hFmJ
aaZ/twgfuK+irMCkjxTXBTbrHhJSoQdNPDR2aX5eJdip0dW+tpARwKR+swvOXybuNftnSGm/SwE6
yVBs4BU8++Wswgg1cR2UC9mogeqLmsL0O49CsW03n+KUi2nmzKXCS79UAtO43O0cIZBZp0f6etQ6
v2cstT9KybcRKb+4czzEFNxtLuIwhYgNyritg2A5Ols3mTobIzxzH+XiO1jT86QcoceYIMNt61tp
xBWjW1jBM3F4tz0AEQEAAQ==
device-key-sha3-384: Qxgo3bZNrgZoFsZf2R4bGAGm_1hlRlU5vL3VxWfUdygbc7FaZuW4is9MpEpRO2sk
sign-key-sha3-384:   wrfougkz3Huq2T_KklfnufCC0HzG7bJ9wP99GV0FF-D3QH3eJtuSRlQc2JhrAoh1
timestamp:           2018-04-14T13:58:20.487877Z
~$ 
```

particularly bad is that `device-key` behemoth in the middle of the thing, which breaks _everything_. Even the raw `--assertion` output gets this one better.

`cmd_info.go` has a bunch of helpers you might use to make things simpler. Reach out to me if they're too obscure (they might well be).

Reply:
I knew this was producing nasty output particularly around the device-key, but I tried looking through some of the helpers in cmd_info.go but it wasn't really clear to me how to use any of them without reading through all 748 lines of cmd_info.go, they all seem pretty specific to the machinery of snap info unfortunately...
I pushed this PR in the hope that someone like you would see this in my PR and go "don't be silly and try to do all this yaml formatting code yourself, just use XYZ yamlish helper package that does the right thing everywhere all the time:tm:", but it seems that is not the case :-|

Reply:
Alright, I think the formatting code is good, let me know if what I have now is okay

Reply:
I'll take a detailed look in my morning :-)